File: core/strategies/payoff.py
"""Payoff curve computation at expiration.

Domain:    Strategy Analysis — Payoff
Contracts:
  - payoff_at_expiration(legs, prices) -> np.ndarray
  - net_premium(legs) -> float
  - find_breakevens(prices, pnl) -> list[float]
Dependencies UPWARD:
  - numpy
Dependencies DOWNWARD:
  - services.strategy_service, core.strategies.greeks
"""

from __future__ import annotations

import numpy as np

from core.strategies.models import Leg


def payoff_at_expiration(legs: list[Leg], prices: np.ndarray) -> np.ndarray:
    """Per-share P&L at expiration."""
    pnl = np.zeros_like(prices, dtype=float)
    for leg in legs:
        if leg.option_type == "call":
            intrinsic = np.maximum(prices - leg.strike, 0.0)
        else:
            intrinsic = np.maximum(leg.strike - prices, 0.0)
        pnl += leg.sign * leg.qty * (intrinsic - leg.premium)
    return pnl


def net_premium(legs: list[Leg]) -> float:
    """Negative = debit paid; positive = credit received."""
    return float(sum(-leg.sign * leg.qty * leg.premium for leg in legs))


def find_breakevens(prices: np.ndarray, pnl: np.ndarray) -> list[float]:
    """Return all prices where P&L crosses zero (linear interpolation)."""
    breakevens: list[float] = []
    sign = np.sign(pnl)
    sign_changes = np.where(np.diff(sign) != 0)[0]
    for i in sign_changes:
        x0, x1 = prices[i], prices[i + 1]
        y0, y1 = pnl[i], pnl[i + 1]
        if y1 == y0:
            continue
        be = x0 - y0 * (x1 - x0) / (y1 - y0)
        breakevens.append(round(float(be), 4))
    return breakevens

File: site/static/sim/payoff.js
// payoff.js — expiry payoff, net premium, breakevens.
//
// Port of core/strategies/payoff.py. Per-share (no contract multiplier) to keep
// numeric parity with the backend; apply any multiplier only at the display
// layer.
//
// A `leg` is a plain object: { side:'long'|'short', optionType:'call'|'put',
// strike, premium, qty=1, dte=30, iv=0.25 }. sign: long = +1, short = -1.

function _sign(leg) {
  return leg.side === 'long' ? 1 : -1;
}

export function payoffAtExpiration(legs, prices) {
  const pnl = prices.map(() => 0);
  for (const leg of legs) {
    const s = _sign(leg);
    const q = leg.qty ?? 1;
    for (let i = 0; i < prices.length; i++) {
      const S = prices[i];
      const intrinsic =
        leg.optionType === 'call' ? Math.max(S - leg.strike, 0) : Math.max(leg.strike - S, 0);
      pnl[i] += s * q * (intrinsic - leg.premium);
    }
  }
  return pnl;
}

export function netPremium(legs) {
  let total = 0;
  for (const leg of legs) {
    total += -_sign(leg) * (leg.qty ?? 1) * leg.premium;
  }
  return total; // negative = debit paid, positive = credit received
}

// Linear-interpolation breakevens, mirroring find_breakevens().
export function findBreakevens(prices, pnl) {
  const out = [];
  for (let i = 0; i < pnl.length - 1; i++) {
    const y0 = pnl[i];
    const y1 = pnl[i + 1];
    if (Math.sign(y1) === Math.sign(y0)) continue; // np.diff(sign) != 0
    if (y1 === y0) continue;
    const x0 = prices[i];
    const x1 = prices[i + 1];
    const be = x0 - y0 * (x1 - x0) / (y1 - y0);
    out.push(Math.round(be * 1e4) / 1e4);
  }
  return out;
}
